custom-types-quick-guide.md

```diff
@@ -0,0 +1,245 @@
+
+These need to be copied to the Hazelcast project under `com.hazelcast.client.impl.protocol.codec`
+for the `DynamicConfigAddUsefulConfigCodec.java` and under
+`com.hazelcast.client.impl.protocol.codec.custom` for `ServiceAConfigCodec.java` and resp.
+`ServiceBConfigCodec.java`.
+
+At this point you are pretty much done with `ServiceAConfig` as the generated code will
+automatically map the protocol type to your domain type. However, for `ServiceBConfig` you need to
+deserialize `addressParser` `Data` field into an actual `AddressParser` and then manually
+instantiate your `ServiceBConfig` type.
+
+For many examples of how this is done I recommend looking at
+`ClientDynamicClusterConfig#addWanReplicationConfig` which shows mapping of the domain types to
+their respective holder types on the way out to the server; and `AddWanReplicationConfigTask` which
+shows the mapping of holder types to their respective configuration types. These can be found in the
+Hazelcast project [here](https://github.com/hazelcast/hazelcast). This example covers most scenarios
+you would encounter.
+
+## Binary Compatibility Tests
+
+The generator produces binary compatibility tests. Currently this step will be broken if you have
+followed the steps. To fix the errors add the following to `binary/__init__.py`'s
+`CustomConfigTypes` list:
+
+```python
+    "ServiceAConfig",
+    "ServiceBConfig"
+```
+
+Now also add the following to `binary/util.py`'s `reference_objects_dict`:
+
+```python
+    'ServiceAConfig': 'aServiceAConfig',
+    'ServiceBConfig': 'aServiceBConfig',
+```
+
+These are the names to which the generated tests will refer to an instance of the respective type.
+
+You will need to copy the artifacts of the API version you just added to, in our case 2.7 to the
+Hazelcast project under the paths listed.
+
+- `output/java/hazelcast/src/test/resources/2.7.protocol.compatibility.binary`
+- `output/java/hazelcast/src/test/resources/2.7.protocol.compatibility.null.binary`
+- `output/java/hazelcast/src/test/java/com/hazelcast/client/protocol/compatibility/ClientCompatibilityNullTest_2_7.java`
+- `output/java/hazelcast/src/test/java/com/hazelcast/client/protocol/compatibility/ClientCompatibilityTest_2_7.java`
+- `output/java/hazelcast/src/test/java/com/hazelcast/client/protocol/compatibility/MemberCompatibilityNullTest_2_7.java`
+- `output/java/hazelcast/src/test/java/com/hazelcast/client/protocol/compatibility/MemberCompatibilityTest_2_7.java`
+
+If you open the test files up in your IDE you will see some reference errors for `aServiceAConfig`
+and `aServiceBConfig`. For example, if you were to open `ClientCompatibilityTest_2_7.java`. You must
+add these to `com.hazelcast.client.protocol.compatibility.ReferenceObjects` in the Hazelcast
+project. If you need a `String` when creating instances of `ServiceAConfig` or `ServiceBConfig` you
+should use `aString` and likewise for other basic types. You should follow this naming strategy when
```

Comment:
Also as I understand, for any array type it should be a singletonList of that type. Worth adding?

Reply:
I pointed them to another example for more detail so if they need a load more examples of how to create bits they can look there. The main point of this quick doc is to describe the undocumented pieces that are required in a relatively succinct way.